lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs

```diff
@@ -257,6 +277,29 @@ receiveD :: UTxO -> UTxO -> (DeltaUTxO, UTxO)
 receiveD a b = (da, a <> b)
   where da = DeltaUTxO { excluded = mempty, received = b }
 
+--------------------------------------------------------------------------------
+-- Queries
+--------------------------------------------------------------------------------
+
+assetIds :: UTxO -> Set AssetId
+assetIds (UTxO u) = foldMap txOutAssetIds u
+
+txIds :: UTxO -> Set (Hash "Tx")
+txIds (UTxO u) = Set.map (view #inputId) (Map.keysSet u)
+
+--------------------------------------------------------------------------------
+-- Transformations
+--------------------------------------------------------------------------------
+
+mapAssetIds :: (AssetId -> AssetId) -> UTxO -> UTxO
+mapAssetIds f (UTxO u) = UTxO $ Map.map (txOutMapAssetIds f) u
+
+mapTxIds :: (Hash "Tx" -> Hash "Tx") -> UTxO -> UTxO
+mapTxIds f (UTxO u) = UTxO $ Map.mapKeys (over #inputId f) u
```

Comment:
I suspect this doesn't satisfy the following composition property:

```hs
mapTxIds f (mapTxIds g u) == mapTxIds (f . g) u
```

If there's a key collision, then one value must be dropped (since there's no way to combine `TxOut` values). Therefore, we need a consistent way to choose between pairs of `TxOut` values.

Since `TxOut` has an `Ord` instance, we could just choose the "smaller" value (using the `min` function).

Reply:
Fixed in 5b8935a0badd0b6508b5e2655aa27dc1eb0d844a.